paper.md

```diff
@@ -120,15 +120,91 @@ Several hosted CI services like Travis, Circle and Jenkins make continuous integ
+  args: [
+    "--create-dirs",
+    "-Lo data/global.csv",
+    "https://github.com/datasets/co2-fossil-global/raw/master/global.csv"
+  ]
+
+- id: run analysis
+  uses: docker://python:alpine
+  args: [
+    "scripts/get_mean_by_group.py",
+    "data/global.csv", "5"
+  ]
+
+- id: validate results
+  uses: docker://python:alpine
+  args: [
+    "scripts/validate_output.py",
+    "data/global_per_capita_mean.csv"
+  ]
+```
+
+A popper workflow consists of a series of syntactical components called steps, where each step represents a node in the workflow DAG, with a `uses` attribute specifying the required container image. 
+The `uses` attribute can reference Docker images hosted in container image registries; filesystem paths for locally defined container images (Dockerfiles); or publicly accessible GitHub repositories that contain Dockerfiles. 
+The commands or scripts that need to be executed in a container can be defined by the `args` and `runs` attributes. 
+Secrets and environment variables needed by a step can be specified by the `secrets` and `env` attributes respectively for making them available inside the container associated to a step.
+The steps in a workflow are executed sequentially in the order in which they are defined.
 
-### **Resource manager API**
+## Workflow Execution Engine
 
-### **Container engine API and plugins**
+The Popper workflow execution engine is composed of several components which talk to each other during a workflow execution.
+The vital architectural components of the system are described in detail throughout this section.
+The architecture of the Popper workflow engine is shown in @Fig:arch;
+
+### Command Line Interface (CLI)
+
+Besides allowing users to communicate with the workflow runner, the CLI provides with search capabilities so that users can search for existing actions implemented by other researchers and developers;
+allows visualizing workflows by generating DOT diagrams [@dot] like the one shown in @Fig:casestudy;
+generates configuration files for continuous integration systems, e.g. TravisCI, Jenkins, Gitlab-CI, etc. so that users can continuously validate the workflows they implement;
+provides dynamic workflow variable substitution capabilities, among others.
+
+### Workflow Definition and Configuration Parsers
+
+The workflow file and the configuration file are parsed by their respective parser plugins at the initial stages of a workflow execution.
+The parsers are responsible for reading and parsing the YML files into an internal format;
+running syntactic and semantic validation checks;
+normalizing the various attributes and generating a workflow DAG.
+The workflow parser has a pluggable architecture that allows adding support of other workflow languages.
+
+### Workflow Runner
+
+The Workflow runner is in charge of parsing the `.yml` files, i.e. the workflow files and the configuration files through their corresponding parsers and creating an internal representation for it. 
```

Comment:
The workflow runner runs a workflow, whereas the parsing is done by the parsers, so we can remove this from the first sentence. We can say "The runner is in charge of taking the workflow representation and executing it".